pkg/scheduler/controller/shoot/scheduler_control.go

```diff
@@ -218,12 +219,21 @@ func filterSeedsMatchingSeedSelector(cloudProfile *gardencorev1beta1.CloudProfil
 	return matchingSeeds, nil
 }
 
-func getCandidates(shoot *gardencorev1beta1.Shoot, seedList []*gardencorev1beta1.Seed, strategy config.CandidateDeterminationStrategy) ([]*gardencorev1beta1.Seed, error) {
+func getCandidates(cloudProfile *gardencorev1beta1.CloudProfile, shoot *gardencorev1beta1.Shoot, seedList []*gardencorev1beta1.Seed, strategy config.CandidateDeterminationStrategy) ([]*gardencorev1beta1.Seed, error) {
 	var candidates []*gardencorev1beta1.Seed
+	var providers []string
 
+	if cloudProfile.Spec.SeedSelector != nil {
+		providers = cloudProfile.Spec.SeedSelector.Providers
+	}
 	switch {
 	case shoot.Spec.Purpose != nil && *shoot.Spec.Purpose == gardencorev1beta1.ShootPurposeTesting:
 		candidates = determineCandidatesOfSameProvider(seedList, shoot)
+	case strategy == config.Selector:
```

Comment:
Hm, why is it a separate strategy? I think this should be the default behaviour in all strategies. If `.spec.seedSelector.providers[]` was provided in the `CloudProfile` then they will be considered. The strategy is only used later to find the best match. WDYT? As you suggested, it probably makes more sense to move the scheduling strategy out of the gardener-scheduler configuration and into the `CloudProfile`, but let's do this in a separate PR.

Reply:
You are right. Compared to the old base for the pull request, this behaviour has changed. Now the label selector is always matched prior to the concrete strategy. This should be kept and additionally the providers from the optional list should be filtered. I'll change this.

Reply:
Hmm, this will not work, because the other strategies check the type. This check could be removed, because it is would be  filtered already by the `matchProvider` filter. But anyway it cannot be combined with `SameRegion` strategy, because the regions of different providers will probably never match. Therefore this provider filter would be completely without function.

Reply:
OK, good points, agreed that the `SameRegion` strategy doesn't make sense cross-provider.

What do you think about the following:

1. As suggested earlier, don't introduce a new strategy, but adapt the two existing ones:
1. The `SameRegion` prefers seeds having providers of the same type over other providers. If a seed of the same type exists then it chooses one with the same region. If no seed with the same provider exists then it chooses any seed with another matching provider, independent of the region (yes, it's not too intuitive for now, but until the whole region/minimal distance stuff is sorted out this is maybe an acceptable behaviour?)
1. The `MinimalDistance` strategy prefers seeds having providers of the same type and then chooses the one with the smallest distance (lexicographically, for now). If no such seed exists it chooses a seed of another matching provider and, again, finds the best region fit lexicographically.

Reply:
This would counteract the name SameReagion. Instead I left the SameRegion as it is and introcuded a `BestRegion` which is a combination of both. It uses `SameRegion` if such a seed exsists. Otherwise it falls back to `MinimalDistance`.

At the same time I improved the `MinimalDistance` strategy. It splits the region names into two parts. The **base region** and an **orientation** (north, south, east, west, central). The distance is then based on the levenshtein distance of the base region plus an adjustment for mismatching orientations. This assures that two regions differing only by their orientation are nearer to each other than two completely different regions.